File: domain/src/main/java/com/beat/domain/schedule/domain/Schedule.java
package com.beat.domain.schedule.domain;

import com.beat.domain.schedule.exception.ScheduleErrorCode;
import com.beat.global.common.exception.ConflictException;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDateTime;

@Entity
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Schedule {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false)
	private LocalDateTime performanceDate;

	@Column(nullable = false)
	private int totalTicketCount;

	@Column(nullable = false)
	private int soldTicketCount = 0;

	@Column(nullable = false)
	private boolean isBooking = true;

	@Enumerated(EnumType.STRING)
	@Column(nullable = false)
	private ScheduleNumber scheduleNumber;

	@Column(name = "performance_id", nullable = false)
	private Long performanceId;

	@Builder
	private Schedule(LocalDateTime performanceDate, int totalTicketCount, int soldTicketCount, boolean isBooking,
		ScheduleNumber scheduleNumber, Long performanceId) {
		this.performanceDate = performanceDate;
		this.totalTicketCount = totalTicketCount;
		this.soldTicketCount = soldTicketCount;
		this.isBooking = isBooking;
		this.scheduleNumber = scheduleNumber;
		this.performanceId = performanceId;
	}

	public static Schedule create(LocalDateTime performanceDate, int totalTicketCount, int soldTicketCount,
		boolean isBooking, ScheduleNumber scheduleNumber, Long performanceId) {
		return Schedule.builder()
			.performanceDate(performanceDate)
			.totalTicketCount(totalTicketCount)
			.soldTicketCount(soldTicketCount)
			.isBooking(isBooking)
			.scheduleNumber(scheduleNumber)
			.performanceId(performanceId)
			.build();
	}

	public void update(LocalDateTime performanceDate, int totalTicketCount, ScheduleNumber scheduleNumber) {
		this.performanceDate = performanceDate;
		this.totalTicketCount = totalTicketCount;
		this.scheduleNumber = scheduleNumber;
	}

	public void decreaseSoldTicketCount(int count) {
		if (this.soldTicketCount >= count) {
			this.soldTicketCount -= count;
		} else {
			throw new ConflictException(ScheduleErrorCode.EXCESS_TICKET_DELETE);
		}
	}

	public void updateScheduleNumber(ScheduleNumber scheduleNumber) {
		this.scheduleNumber = scheduleNumber;
	}

	public void updateIsBooking(boolean isBooking) {
		this.isBooking = isBooking;
	}
}

File: domain/src/main/kotlin/com/beat/domain/schedule/domain/Schedule.kt
package com.beat.domain.schedule.domain

import com.beat.domain.performance.domain.Performance
import com.beat.domain.schedule.exception.ScheduleErrorCode
import com.beat.global.support.exception.BadRequestException
import com.beat.global.support.exception.ConflictException
import java.time.LocalDateTime

@ConsistentCopyVisibility
data class Schedule private constructor(
    private val scheduleId: Id?,
    private val performanceDate: LocalDateTime,
    private val bookingCloseAt: LocalDateTime,
    private val totalTicketCount: Int,
    private val soldTicketCount: Int,
    private val scheduleNumber: ScheduleNumber,
    private val linkedPerformanceId: Performance.Id,
) {
    fun getId(): Long? = scheduleId?.value

    fun getPerformanceId(): Long = linkedPerformanceId.value

    fun getPerformanceDate(): LocalDateTime = performanceDate

    fun getBookingCloseAt(): LocalDateTime = bookingCloseAt

    fun getTotalTicketCount(): Int = totalTicketCount

    fun getSoldTicketCount(): Int = soldTicketCount

    fun getScheduleNumber(): ScheduleNumber = scheduleNumber

    fun update(
        performanceDate: LocalDateTime,
        bookingCloseAt: LocalDateTime,
        totalTicketCount: Int,
        scheduleNumber: ScheduleNumber,
    ): Schedule {
        validateTicketCounts(totalTicketCount, soldTicketCount)
        validateBookingWindow(performanceDate, bookingCloseAt)

        return copy(
            performanceDate = performanceDate,
            bookingCloseAt = bookingCloseAt,
            totalTicketCount = totalTicketCount,
            scheduleNumber = scheduleNumber
        )
    }

    fun increaseSoldTicketCount(count: Int): Schedule {
        validatePositiveCount(count)

        val updatedSoldTicketCount = soldTicketCount + count
        if (updatedSoldTicketCount > totalTicketCount) {
            throw ConflictException(ScheduleErrorCode.INSUFFICIENT_TICKETS)
        }

        return copy(soldTicketCount = updatedSoldTicketCount)
    }

    fun decreaseSoldTicketCount(count: Int): Schedule {
        validatePositiveCount(count)

        if (soldTicketCount < count) {
            throw ConflictException(ScheduleErrorCode.EXCESS_TICKET_DELETE)
        }
        return copy(soldTicketCount = soldTicketCount - count)
    }

    fun updateScheduleNumber(scheduleNumber: ScheduleNumber): Schedule = copy(scheduleNumber = scheduleNumber)

    fun updateBookingCloseAt(bookingCloseAt: LocalDateTime): Schedule {
        validateBookingWindow(performanceDate, bookingCloseAt)
        return copy(bookingCloseAt = bookingCloseAt)
    }

    private fun validatePositiveCount(count: Int) {
        if (count <= 0) {
            throw BadRequestException(ScheduleErrorCode.INVALID_DATA_FORMAT)
        }
    }

    @JvmInline
    value class Id private constructor(val value: Long) {
        companion object {
            @JvmStatic
            fun from(value: Long): Id = Id(value)

            @JvmStatic
            fun fromNullable(value: Long?): Id? = value?.let(::from)
        }
    }

    companion object {
        @JvmStatic
        fun create(
            performanceDate: LocalDateTime,
            bookingCloseAt: LocalDateTime,
            totalTicketCount: Int,
            scheduleNumber: ScheduleNumber,
            performanceId: Long,
        ): Schedule {
            validateTicketCounts(totalTicketCount, 0)
            validateBookingWindow(performanceDate, bookingCloseAt)

            return Schedule(
                scheduleId = null,
                performanceDate = performanceDate,
                bookingCloseAt = bookingCloseAt,
                totalTicketCount = totalTicketCount,
                soldTicketCount = 0,
                scheduleNumber = scheduleNumber,
                linkedPerformanceId = Performance.Id.from(performanceId)
            )
        }

        @JvmStatic
        fun rehydrate(
            id: Long?,
            performanceDate: LocalDateTime,
            bookingCloseAt: LocalDateTime,
            totalTicketCount: Int,
            soldTicketCount: Int,
            scheduleNumber: ScheduleNumber,
            performanceId: Long,
        ): Schedule {
            validateTicketCounts(totalTicketCount, soldTicketCount)
            validateBookingWindow(performanceDate, bookingCloseAt)

            return Schedule(
                scheduleId = Id.fromNullable(id),
                performanceDate = performanceDate,
                bookingCloseAt = bookingCloseAt,
                totalTicketCount = totalTicketCount,
                soldTicketCount = soldTicketCount,
                scheduleNumber = scheduleNumber,
                linkedPerformanceId = Performance.Id.from(performanceId)
            )
        }

        private fun validateTicketCounts(totalTicketCount: Int, soldTicketCount: Int) {
            if (totalTicketCount < 0 || soldTicketCount < 0 || soldTicketCount > totalTicketCount) {
                throw BadRequestException(ScheduleErrorCode.INVALID_DATA_FORMAT)
            }
        }

        private fun validateBookingWindow(performanceDate: LocalDateTime, bookingCloseAt: LocalDateTime) {
            if (bookingCloseAt.isBefore(performanceDate)) {
                throw BadRequestException(ScheduleErrorCode.INVALID_DATA_FORMAT)
            }
        }
    }
}
